### PR TITLE
[#160235] Upgrade to Rails 7

### PR DESCRIPTION
**Pull request by lautarob**
# Release Notes

Upgrades Rails from 6.1.7 to 7.0.8

# Defaults to still implement

`cookies_serializer` has not been updated in this PR. A separate PR has been created for that: https://github.com/tablexi/nucore-open/pull/3691

# New defaults in application.rb

`cache_format_version` and `disable_to_s_conversion` have been set in `application.rb`, as indicated by the documentation.

# Testing

See [this list](https://www.notion.so/txi/Major-application-capabilities-for-testing-762abdf40ba047f6b4fc6ddc840fd5ea?pvs=4) in notion.
Make sure NU can access the card reader page:
https://nucore-staging.northwestern.edu/facilities/all/users/23782/card_number/edit

### CODE REVIEW

**Comment by giladshanan** [comment]
@jossim I think there's just 1 failing spec remaining - but it may be a tricky one.  Have you looked at deprecations or upgrading the defaults?

File: config/boot.rb [hunk 1]
```diff
@@ -1,6 +1,4 @@
-# frozen_string_literal: true
+ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../Gemfile", __dir__)
```
Comment (giladshanan) [inline config/boot.rb:1]:
```suggestion
# frozen_string_literal: true

ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../Gemfile", __dir__)
```

File: config/override_locales/en.yml [hunk 2]
```diff
@@ -1,13 +1,8 @@
...
 en:
-  # app_name: NUcore
-  # institution_name: Northwestern
-
-  # shared:
-  #   footer:
-  #     copyright_html: "&copy; Copyright 2011%{to_date} Northwestern University"
-  #     logo_alt: "Northwestern University Logo"
-
+  views:
```
Comment (giladshanan) [inline config/override_locales/en.yml:2]: @lautarob no changes should be committed for this file - please revert them.

**Review by giladshanan — COMMENTED** [review]
@lautarob just a couple small changes

File: config/initializers/new_framework_defaults_7_0.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+# `stylesheet_link_tag` view helper will not render the media attribute by default.
+Rails.application.config.action_view.apply_stylesheet_media_default = false
+
+# Change the digest class for the key generators to `OpenSSL::Digest::SHA256`.
+# Changing this default means invalidate all encrypted messages generated by
+# your application and, all the encrypted cookies. Only change this after you
+# rotated all the messages using the key rotator.
+#
+# See upgrading guide for more information on how to build a rotator.
```
Comment (giladshanan) [inline config/initializers/new_framework_defaults_7_0.rb:24]: @jossim looks like we need to implement a rotator for this - have you looked into that?

File: config/initializers/new_framework_defaults_7_0.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+# Active Storage `has_many_attached` relationships will default to replacing the current collection instead of appending to it.
+# Thus, to support submitting an empty collection, the `file_field` helper will render an hidden field `include_hidden` by default when `multiple_file_field_include_hidden` is set to `true`.
+# See https://guides.rubyonrails.org/configuring.html#config-active-storage-multiple-file-field-include-hidden for more information.
+Rails.application.config.active_storage.multiple_file_field_include_hidden = true
+
+# ** Please read carefully, this must be configured in config/application.rb (NOT this file) **
+# Disables the deprecated #to_s override in some Ruby core classes
+# See https://guides.rubyonrails.org/configuring.html#config-active-support-disable-to-s-conversion for more information.
+# config.active_support.disable_to_s_conversion = true
```
Comment (giladshanan) [inline config/initializers/new_framework_defaults_7_0.rb:143]: I think this is duplicated above...

File: config/initializers/new_framework_defaults_7_0.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+# ** Please read carefully, this must be configured in config/application.rb **
+# Change the format of the cache entry.
+# Changing this default means that all new cache entries added to the cache
+# will have a different format that is not supported by Rails 6.1 applications.
+# Only change this value after your application is fully deployed to Rails 7.0
+# and you have no plans to rollback.
+# When you're ready to change format, add this to `config/application.rb` (NOT this file):
+#  config.active_support.cache_format_version = 7.0
+
+
```
Comment (giladshanan) [inline config/initializers/new_framework_defaults_7_0.rb:105]: Looks like this is a duplicate

File: spec/teaspoon_env.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # frozen_string_literal: true
+require 'engine_manager'
```
Comment (giladshanan) [inline spec/teaspoon_env.rb:2]:
```suggestion
require "engine_manager"
```

File: vendor/engines/secure_rooms/lib/secure_rooms/engine.rb [hunk 7]
```diff
@@ -43,12 +43,8 @@ class Engine < ::Rails::Engine
 
       ViewHook.add_hook "search.results_table", "extra_headers", "search/extra_headers"
       ViewHook.add_hook "search.results_table", "extra_columns", "search/extra_columns"
-    end
 
-    initializer "secure_rooms.action_controller" do
-      ActiveSupport.on_load :action_controller do
-        helper SecureRooms::SecureRoomsHelper
-      end
+      ApplicationController.helper SecureRooms::SecureRoomsHelper
```
Comment (giladshanan) [inline vendor/engines/secure_rooms/lib/secure_rooms/engine.rb:47]: We should make a note of how to test this on NU stage

**Review by giladshanan — COMMENTED** [review]
@jossim left a few comments

File: config/initializers/new_framework_defaults_7_0.rb (hunk 5 above)
Comment (jossim) [reply config/initializers/new_framework_defaults_7_0.rb:105]: I did copy this and the one at the end of the file over to `application.rb`, if that's what you're saying

Comment (giladshanan) [reply config/initializers/new_framework_defaults_7_0.rb:105]: @jossim I see - why did you put the changes there instead of using the defaults file?

Comment (jossim) [reply config/initializers/new_framework_defaults_7_0.rb:105]: @giladshanan the documentation indicated that changes to `disable_to_s_conversion` and `cache_format_version` should go in `application.rb`

File: vendor/engines/secure_rooms/lib/secure_rooms/engine.rb (hunk 7 above)
Comment (giladshanan) [reply vendor/engines/secure_rooms/lib/secure_rooms/engine.rb:47]: https://nucore-staging.northwestern.edu/facilities/all/users/23782/card_number/edit

**Comment by giladshanan** [comment]
@jossim the Tasks section of the PR description needs an update, and there is a merge conflict to handle in schema.rb

File: db/schema.rb [hunk 8]
```diff
@@ -459,8 +458,8 @@
     t.integer "price_group_id", null: false
     t.integer "schedule_rule_id", null: false
     t.decimal "discount_percent", precision: 10, null: false
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
```
Comment (giladshanan) [inline db/schema.rb:463]: This seems odd

Comment (giladshanan) [reply db/schema.rb:463]: Seems correct: https://github.com/rails/rails/issues/44571#issuecomment-1054758649

**Review by giladshanan — APPROVED** [review]

**Pull request merged**